test/integration/controller_instance_test.go

```diff
@@ -992,3 +1055,470 @@ func TestCreateServiceInstanceWithProvisionFailure(t *testing.T) {
+	key := osb.OperationKey(testOperation)
+
+	cases := []struct {
+		name                         string
+		skipVerifyingInstanceSuccess bool
+		binding                      *v1beta1.ServiceBinding
+		setup                        func(*controllerTest)
+		testFunction                 func(t *controllerTest)
+	}{
+		{
+			name:    "synchronous deprovision",
+			binding: getTestBinding(),
+			setup: func(ct *controllerTest) {
+				ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
+					Response: &osb.DeprovisionResponse{},
+				}
+			},
+		},
+		{
+			name: "synchronous deprovision, no binding",
+			setup: func(ct *controllerTest) {
+				ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
+					Response: &osb.DeprovisionResponse{},
+				}
+			},
+		},
+		{
+			name:    "asynchronous deprovision with operation key",
+			binding: getTestBinding(),
+			setup: func(ct *controllerTest) {
+				ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
+					Response: &osb.DeprovisionResponse{
+						Async:        true,
+						OperationKey: &key,
+					},
+				}
+			},
+		},
+		{
+			name: "asynchronous deprovision with operation key, no binding",
+			setup: func(ct *controllerTest) {
+				ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
+					Response: &osb.DeprovisionResponse{
+						Async:        true,
+						OperationKey: &key,
+					},
+				}
+			},
+		},
+		{
+			name:    "asynchronous deprovision without operation key",
+			binding: getTestBinding(),
+			setup: func(ct *controllerTest) {
+				ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
+					Response: &osb.DeprovisionResponse{
+						Async: true,
+					},
+				}
+			},
+		},
+		{
+			name: "asynchronous deprovision without operation key, no binding",
+			setup: func(ct *controllerTest) {
+				ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
+					Response: &osb.DeprovisionResponse{
+						Async: true,
+					},
+				}
+			},
+		},
+		{
+			name:    "deprovision instance with binding not deleted",
+			binding: getTestBinding(),
+			setup: func(ct *controllerTest) {
+				ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
+					Response: &osb.DeprovisionResponse{},
+				}
+			},
+			testFunction: func(ct *controllerTest) {
+				if err := ct.client.ServiceInstances(ct.instance.Namespace).Delete(ct.instance.Name, &metav1.DeleteOptions{}); err != nil {
+					ct.t.Fatalf("instance delete should have been accepted: %v", err)
+				}
+
+				condition := v1beta1.ServiceInstanceCondition{
+					Type:   v1beta1.ServiceInstanceConditionReady,
+					Status: v1beta1.ConditionFalse,
+					Reason: "DeprovisionBlockedByExistingCredentials",
+				}
+				if err := util.WaitForInstanceCondition(ct.client, testNamespace, testInstanceName, condition); err != nil {
+					ct.t.Fatalf("error waiting for instance condition: %v", err)
+				}
+			},
+		},
+		{
+			name: "deprovision instance during in progress provision",
```

Comment:
This test is not actually testing a deprovision during an in-progress provision. The ServiceInstance is fully provisioned before it is deleted.

Reply:
Changed the wording. I struggled to find a way to actually do a deprovision with the provision still in progress.